File: Tests/BridgeJSRuntimeTests/DefaultArgumentTests.swift
import XCTest
import JavaScriptKit

@JSClass struct DefaultArgumentImports {
    @JSFunction static func runJsDefaultArgumentTests() throws(JSException)
}

final class DefaultArgumentTests: XCTestCase {
    func testRunJsDefaultArgumentTests() throws {
        try DefaultArgumentImports.runJsDefaultArgumentTests()
    }
}

// MARK: - Default Parameters

@JS enum DefaultArgumentExports {
    @JS static func testStringDefault(message: String = "Hello World") -> String {
        return message
    }

    @JS static func testIntDefault(count: Int = 42) -> Int {
        return count
    }

    @JS static func testBoolDefault(flag: Bool = true) -> Bool {
        return flag
    }

    @JS static func testOptionalDefault(name: String? = nil) -> String? {
        return name
    }

    @JS static func testMultipleDefaults(
        title: String = "Default Title",
        count: Int = -10,
        enabled: Bool = false
    ) -> String {
        return "\(title): \(count) (\(enabled))"
    }

    @JS static func testSimpleEnumDefault(status: Status = .success) -> Status {
        return status
    }

    @JS static func testDirectionDefault(direction: Direction = .north) -> Direction {
        return direction
    }

    @JS static func testRawStringEnumDefault(theme: Theme = .light) -> Theme {
        return theme
    }

    @JS static func testComplexInit(greeter: Greeter = Greeter(name: "DefaultGreeter")) -> String {
        return greeter.greet()
    }

    @JS static func testEmptyInit(_ object: StaticPropertyHolder = StaticPropertyHolder()) -> StaticPropertyHolder {
        return object
    }

    @JS static func createConstructorDefaults(
        name: String = "Default",
        count: Int = 42,
        enabled: Bool = true,
        status: Status = .success,
        tag: String? = nil
    ) -> DefaultArgumentConstructorDefaults {
        return DefaultArgumentConstructorDefaults(
            name: name,
            count: count,
            enabled: enabled,
            status: status,
            tag: tag
        )
    }

    @JS static func describeConstructorDefaults(
        _ value: DefaultArgumentConstructorDefaults
    ) -> String {
        return value.describe()
    }

    @JS static func arrayWithDefault(_ values: [Int] = [1, 2, 3]) -> Int {
        return values.reduce(0, +)
    }

    @JS static func arrayWithOptionalDefault(_ values: [Int]? = nil) -> Int {
        return values?.reduce(0, +) ?? -1
    }

    @JS static func arrayMixedDefaults(
        prefix: String = "Sum",
        values: [Int] = [10, 20],
        suffix: String = "!"
    ) -> String {
        return "\(prefix): \(values.reduce(0, +))\(suffix)"
    }
}

@JS class DefaultArgumentConstructorDefaults {
    @JS var name: String
    @JS var count: Int
    @JS var enabled: Bool
    @JS var status: Status
    @JS var tag: String?

    @JS init(
        name: String = "Default",
        count: Int = 42,
        enabled: Bool = true,
        status: Status = .success,
        tag: String? = nil
    ) {
        self.name = name
        self.count = count
        self.enabled = enabled
        self.status = status
        self.tag = tag
    }

    @JS func describe() -> String {
        let tagStr = tag ?? "nil"
        let statusStr: String
        switch status {
        case .loading: statusStr = "loading"
        case .success: statusStr = "success"
        case .error: statusStr = "error"
        }
        return "\(name):\(count):\(enabled):\(statusStr):\(tagStr)"
    }
}

File: Tests/BridgeJSRuntimeTests/JavaScript/DefaultArgumentTests.mjs
// @ts-check

import assert from 'node:assert';

/**
 * @returns {import('../../../.build/plugins/PackageToJS/outputs/PackageTests/bridge-js.d.ts').Imports["DefaultArgumentImports"]}
 */
export function getImports(importsContext) {
    return {
        runJsDefaultArgumentTests: () => {
            const exports = importsContext.getExports();
            if (!exports) { throw new Error("No exports!?"); }
            runJsDefaultArgumentTests(exports);
        },
    };
}

/**
 * Default argument bridging coverage for BridgeJS runtime tests.
 * @param {import('../../../.build/plugins/PackageToJS/outputs/PackageTests/bridge-js.d.ts').Exports} rootExports
 */
export function runJsDefaultArgumentTests(rootExports) {
    const exports = rootExports.DefaultArgumentExports;
    const { Status, Direction, Theme } = rootExports;

    assert.equal(exports.testStringDefault(), "Hello World");
    assert.equal(exports.testStringDefault("Custom Message"), "Custom Message");

    assert.equal(exports.testIntDefault(), 42);
    assert.equal(exports.testIntDefault(100), 100);

    assert.equal(exports.testBoolDefault(), true);
    assert.equal(exports.testBoolDefault(false), false);

    assert.equal(exports.testOptionalDefault(), null);
    assert.equal(exports.testOptionalDefault("Test"), "Test");

    assert.equal(exports.testMultipleDefaults(), "Default Title: -10 (false)");
    assert.equal(exports.testMultipleDefaults("Custom"), "Custom: -10 (false)");
    assert.equal(exports.testMultipleDefaults("Custom", 5), "Custom: 5 (false)");
    assert.equal(exports.testMultipleDefaults("Custom", undefined, true), "Custom: -10 (true)");
    assert.equal(exports.testMultipleDefaults("Custom", 5, true), "Custom: 5 (true)");

    assert.equal(exports.testSimpleEnumDefault(), Status.Success);
    assert.equal(exports.testSimpleEnumDefault(Status.Loading), Status.Loading);

    assert.equal(exports.testDirectionDefault(), Direction.North);
    assert.equal(exports.testDirectionDefault(Direction.South), Direction.South);

    assert.equal(exports.testRawStringEnumDefault(), Theme.Light);
    assert.equal(exports.testRawStringEnumDefault(Theme.Dark), Theme.Dark);

    const holder = exports.testEmptyInit();
    assert.notEqual(holder, null);
    holder.release();

    const customHolder = new rootExports.StaticPropertyHolder();
    assert.deepEqual(exports.testEmptyInit(customHolder), customHolder);
    customHolder.release();

    assert.equal(exports.testComplexInit(), "Hello, DefaultGreeter!");
    const customGreeter = new rootExports.Greeter("CustomName");
    assert.equal(exports.testComplexInit(customGreeter), "Hello, CustomName!");
    customGreeter.release();

    const cd1 = exports.createConstructorDefaults();
    assert.equal(exports.describeConstructorDefaults(cd1), "Default:42:true:success:nil");
    cd1.release();

    const cd2 = exports.createConstructorDefaults("Custom");
    assert.equal(exports.describeConstructorDefaults(cd2), "Custom:42:true:success:nil");
    cd2.release();

    const cd3 = exports.createConstructorDefaults("Custom", 100);
    assert.equal(exports.describeConstructorDefaults(cd3), "Custom:100:true:success:nil");
    cd3.release();

    const cd4 = exports.createConstructorDefaults("Custom", undefined, false);
    assert.equal(exports.describeConstructorDefaults(cd4), "Custom:42:false:success:nil");
    cd4.release();

    const cd5 = exports.createConstructorDefaults("Test", 99, false, Status.Loading);
    assert.equal(exports.describeConstructorDefaults(cd5), "Test:99:false:loading:nil");
    cd5.release();

    assert.equal(exports.arrayWithDefault(), 6);
    assert.equal(exports.arrayWithDefault([10, 20]), 30);
    assert.equal(exports.arrayWithOptionalDefault(), -1);
    assert.equal(exports.arrayWithOptionalDefault(null), -1);
    assert.equal(exports.arrayWithOptionalDefault([5, 5]), 10);
    assert.equal(exports.arrayMixedDefaults(), "Sum: 30!");
    assert.equal(exports.arrayMixedDefaults("Total"), "Total: 30!");
    assert.equal(exports.arrayMixedDefaults("Total", [1, 2, 3]), "Total: 6!");
    assert.equal(exports.arrayMixedDefaults("Val", [100], "?"), "Val: 100?");
    assert.equal(exports.arrayMixedDefaults(undefined, [5, 5]), "Sum: 10!");
    assert.equal(exports.arrayMixedDefaults(undefined, undefined, "?"), "Sum: 30?");
}
